File: ios/TacticalMaps/Waypoints/SymbolControlsCard.swift
import SwiftUI
import CoreLocation

/// Floating compact card that appears when the user taps any waypoint
/// annotation on the map. Surfaces the actions that make sense for
/// the kind:
///   - All kinds: live preview, name, "Move to Crosshair", close.
///   - Tactical control measures: rotation + size sliders.
///   - Military / generic: just move.
///
/// Designed to sit just above the bottom safe-area inset so it doesn't
/// overlap with the "Centre on My Location" pill. Tap-outside dismissal
/// is handled by `ContentView` — this view only renders.
struct SymbolControlsCard: View {
    @ObservedObject var waypointStore: WaypointStore
    /// Shared layer model — used to render and pick the waypoint's layer.
    @ObservedObject var drawingStore: DrawingStore
    /// Map VM exposes the current crosshair coordinate (camera centre)
    /// for the "Move to Crosshair" action.
    @ObservedObject var mapVM: MapViewModel
    /// ID of the waypoint we're editing. The view re-resolves the
    /// current Waypoint from the store on every redraw so changes
    /// persist immediately and the preview stays in sync.
    let waypointID: UUID
    let onDismiss: () -> Void

    @State private var showDeleteConfirm: Bool = false
    /// Tapping the name in the header presents the full edit sheet so the
    /// user can change the kind (e.g. swap Ambush → Block) without
    /// re-creating the waypoint.
    @State private var showingEdit: Bool = false

    var body: some View {
        if let wp = waypointStore.waypoints.first(where: { $0.id == waypointID }) {
            card(for: wp)
        }
    }

    private func card(for wp: Waypoint) -> some View {
        VStack(spacing: 8) {
            header(for: wp)

            // Rotation + width/height live only for tactical control
            // measures — military symbols don't have orientation or
            // per-instance size in the model. No dividers between
            // sections — the icons and spacing carry enough structure.
            if case .controlMeasure = wp.kind {
                colorRow(for: wp)
                rotationRow(for: wp)
                widthRow(for: wp)
                heightRow(for: wp)
            }

            actionRow(for: wp)
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 10)
        .background(.regularMaterial, in: RoundedRectangle(cornerRadius: 16, style: .continuous))
        .overlay(
            RoundedRectangle(cornerRadius: 16, style: .continuous)
                .stroke(.white.opacity(0.15), lineWidth: 0.5)
        )
        .shadow(color: .black.opacity(0.25), radius: 10, y: 4)
        .alert("Delete symbol?", isPresented: $showDeleteConfirm) {
            Button("Delete", role: .destructive) {
                if let wp = waypointStore.waypoints.first(where: { $0.id == waypointID }) {
                    waypointStore.remove(wp)
                }
                onDismiss()
            }
            Button("Cancel", role: .cancel) { }
        } message: {
            let name = waypointStore.waypoints
                .first(where: { $0.id == waypointID })?.name ?? "this symbol"
            Text("This will permanently remove “\(name)”.")
        }
        .sheet(isPresented: $showingEdit) {
            if let wp = waypointStore.waypoints.first(where: { $0.id == waypointID }) {
                WaypointEditSheet(
                    waypointStore: waypointStore,
                    original: wp,
                    defaultCoordinate: wp.coordinate
                )
            }
        }
    }

    // MARK: Rows

    private func header(for wp: Waypoint) -> some View {
        // Compact one-line header: small icon + name (or name + kind
        // muted, if they differ) + close. The previous two-line layout
        // wasted vertical space and the subtitle was usually the same
        // string as the name (we auto-fill blank names from the kind's
        // display name).
        let kindLabel = wp.kind.displayName
        let showKindSuffix = wp.name != kindLabel
        return HStack(spacing: 10) {
            // Military unit symbols carry an echelon (dots / bars / X)
            // sitting *above* the frame plus a function glyph inside —
            // a 22pt icon in a 28pt tile compressed all of that into
            // an unreadable blob. Give units a roomier tile; control
            // measures + generic markers stay compact.
            let isUnit = wp.kind.militarySpec != nil
            let tile: CGFloat = isUnit ? 44 : 28
            let inner: CGFloat = isUnit ? 38 : 22
            WaypointKindIcon(
                kind: wp.kind,
                size: inner,
                rotation: wp.kind.controlMeasure == nil ? 0 : wp.rotation,
                taskColor: wp.taskColor
            )
            .frame(width: tile, height: tile)
            // White background so the (mostly black) symbols stay
            // legible against the translucent material card.
            .background(
                RoundedRectangle(cornerRadius: 6, style: .continuous)
                    .fill(Color.white)
            )

            Button {
                showingEdit = true
            } label: {
                HStack(spacing: 6) {
                    VStack(alignment: .leading, spacing: 1) {
                        Text(wp.name)
                            .font(.subheadline.weight(.semibold))
                            .foregroundStyle(.primary)
                            .lineLimit(1)
                        if showKindSuffix {
                            Text(kindLabel)
                                .font(.caption2)
                                .foregroundStyle(.secondary)
                                .lineLimit(1)
                        }
                    }
                    Image(systemName: "pencil")
                        .font(.caption2.weight(.semibold))
                        .foregroundStyle(.tertiary)
                }
                .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityHint("Edit symbol type")
            Spacer(minLength: 4)
            Button(action: onDismiss) {
                Image(systemName: "xmark.circle.fill")
                    .font(.body)
                    .symbolRenderingMode(.hierarchical)
                    .foregroundStyle(.secondary)
            }
            .buttonStyle(.plain)
            .accessibilityLabel("Close symbol controls")
        }
    }

    /// Compact layer pill — colour swatch + name + item count, opens a menu
    /// to reassign.
    private func layerPill(for wp: Waypoint) -> some View {
        let current = drawingStore.layer(id: wp.layerID) ?? drawingStore.layers.first
        return Menu {
            ForEach(drawingStore.layers) { layer in
                let count = layerItemCount(layer)
                Button {
                    var updated = wp
                    updated.layerID = layer.id
                    waypointStore.update(updated)
                } label: {
                    Label("\(layer.name) (\(count))",
                          systemImage: layer.id == current?.id
                              ? "largecircle.fill.circle"
                              : "circle.fill")
                }
                .tint(Color(hex: layer.defaultColorHex))
            }
        } label: {
            HStack(spacing: 6) {
                Image(systemName: "square.stack.3d.up.fill")
                    .font(.footnote.weight(.semibold))
                    .foregroundStyle(Color(hex: current?.defaultColorHex ?? "#888888"))
                Text(current.map { "\($0.name) (\(layerItemCount($0)))" } ?? "—")
                    .font(.footnote.weight(.semibold))
                    .foregroundStyle(.primary)
                    .lineLimit(1)
                    .truncationMode(.tail)
            }
            .frame(maxWidth: 140)
            .padding(.horizontal, 10)
            .padding(.vertical, 6)
            .background(.white.opacity(0.10), in: Capsule())
        }
        .buttonStyle(.plain)
        .accessibilityLabel("Layer")
        .accessibilityValue(current?.name ?? "")
    }

    /// Count of drawings + waypoints assigned to a layer.
    private func layerItemCount(_ layer: DrawingLayer) -> Int {
        let drawings = drawingStore.shapes(in: layer.id).count
        let waypoints = waypointStore.waypoints.filter { $0.layerID == layer.id }.count
        return drawings + waypoints
    }

    /// Five-swatch colour picker for the task graphic. Black is the
    /// default; the others follow the APP-6 affiliation palette
    /// (blue = friendly, red = hostile, green = neutral, yellow =
    /// unknown). Mirrors Android's ControlMeasureControls colour row.
    private func colorRow(for wp: Waypoint) -> some View {
        HStack(spacing: 8) {
            Image(systemName: "paintpalette")
                .font(.footnote)
                .foregroundStyle(.secondary)
                .frame(width: 16)
            HStack(spacing: 10) {
                ForEach(TaskColor.allCases, id: \.self) { tc in
                    Button {
                        var updated = wp
                        updated.taskColor = tc
                        waypointStore.update(updated)
                    } label: {
                        Circle()
                            .fill(tc.color)
                            .frame(width: 28, height: 28)
                            // White hairline so black/dark swatches read on
                            // the translucent dark card.
                            .overlay(Circle().strokeBorder(.white.opacity(0.7), lineWidth: 1))
                            // Accent ring marks the current selection.
                            .overlay(Circle().strokeBorder(
                                Color.accentColor,
                                lineWidth: wp.taskColor == tc ? 3 : 0))
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel(tc.label)
                    .accessibilityAddTraits(wp.taskColor == tc ? [.isSelected] : [])
                }
            }
            Spacer(minLength: 0)
        }
    }

    private func rotationRow(for wp: Waypoint) -> some View {
        sliderRow(
            icon: "arrow.clockwise.circle",
            title: "Rotation",
            valueLabel: "\(Int(wp.rotation.rounded()))°",
            value: Binding(
                get: { wp.rotation },
                set: { newValue in
                    var updated = wp
                    updated.rotation = newValue
                    waypointStore.update(updated)
                }
            ),
            range: 0...360,
            step: 1,
            resetTo: 0
        )
    }

    private func widthRow(for wp: Waypoint) -> some View {
        sliderRow(
            icon: "arrow.left.and.right.circle",
            title: "Width",
            valueLabel: String(format: "%.2f×", wp.scaleX),
            value: Binding(
                get: { wp.scaleX },
                set: { newValue in
                    var updated = wp
                    updated.scaleX = newValue
                    waypointStore.update(updated)
                }
            ),
            range: 0.1...20.0,
            step: 0.1,
            resetTo: 1.0
        )
    }

    private func heightRow(for wp: Waypoint) -> some View {
        sliderRow(
            icon: "arrow.up.and.down.circle",
            title: "Height",
            valueLabel: String(format: "%.2f×", wp.scaleY),
            value: Binding(
                get: { wp.scaleY },
                set: { newValue in
                    var updated = wp
                    updated.scaleY = newValue
                    waypointStore.update(updated)
                }
            ),
            range: 0.1...20.0,
            step: 0.1,
            resetTo: 1.0
        )
    }

    /// Move + Delete row. Move snaps the waypoint to the current map
    /// centre (where the crosshair sits) — long-press-drag on the map
    /// itself is also supported. Delete shows a confirm alert before
    /// removing the waypoint, then dismisses the card.
    private func actionRow(for wp: Waypoint) -> some View {
        HStack(spacing: 8) {
            layerPill(for: wp)
            Button {
                var updated = wp
                updated.latitude  = mapVM.cameraCentre.latitude
                updated.longitude = mapVM.cameraCentre.longitude
                waypointStore.update(updated)
            } label: {
                Label {
                    Text("Move to Crosshair")
                        .font(.footnote.weight(.semibold))
                        .lineLimit(1)
                } icon: {
                    Image(systemName: "scope")
                        .font(.footnote)
                }
                .foregroundStyle(.white)
                .frame(maxWidth: .infinity)
                .padding(.vertical, 7)
                .background(Color.accentColor.opacity(0.85),
                            in: RoundedRectangle(cornerRadius: 8, style: .continuous))
            }
            .buttonStyle(.plain)
            .accessibilityHint("Pan the map first so the crosshair is at the new location, then tap.")

            Button {
                showDeleteConfirm = true
            } label: {
                Image(systemName: "trash")
                    .font(.footnote.weight(.semibold))
                    .foregroundStyle(.white)
                    .frame(width: 36, height: 30)
                    .background(Color.red.opacity(0.85),
                                in: RoundedRectangle(cornerRadius: 8, style: .continuous))
            }
            .buttonStyle(.plain)
            .accessibilityLabel("Delete symbol")
        }
    }

    // MARK: Slider primitive

    /// One-line slider: `[icon] [——slider——] [value] [reset]`. The
    /// `title` is used only for the reset button's accessibility
    /// label — the icon visually conveys what's being adjusted, and
    /// dropping the redundant text label saves a whole row per slider.
    private func sliderRow(icon: String,
                           title: String,
                           valueLabel: String,
                           value: Binding<Double>,
                           range: ClosedRange<Double>,
                           step: Double,
                           resetTo defaultValue: Double) -> some View {
        HStack(spacing: 8) {
            Image(systemName: icon)
                .font(.footnote)
                .foregroundStyle(.secondary)
                .frame(width: 16)
            Slider(value: value, in: range, step: step)
            Text(valueLabel)
                .font(.caption.monospacedDigit())
                .foregroundStyle(.secondary)
                .frame(minWidth: 44, alignment: .trailing)
            Button {
                value.wrappedValue = defaultValue
            } label: {
                Image(systemName: "arrow.counterclockwise")
                    .font(.caption2.weight(.semibold))
                    .frame(width: 22, height: 22)
                    .background(.tint.opacity(0.15), in: Circle())
            }
            .buttonStyle(.plain)
            .accessibilityLabel("Reset \(title)")
        }
    }
}

File: android/app/src/main/java/com/tacticalmaps/map/SymbolControlsCard.kt
package com.tacticalmaps.map

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material.icons.filled.Flag
import androidx.compose.material.icons.filled.GpsFixed
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Slider
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.tacticalmaps.drawings.DrawingLayer
import com.tacticalmaps.waypoints.TaskColor
import com.tacticalmaps.waypoints.Waypoint
import com.tacticalmaps.waypoints.WaypointKind
import com.tacticalmaps.waypoints.WaypointStore

/**
 * Floating controls card for a tapped waypoint. Phase 2 adds the
 * Android-side APP-6C controls plus tactical task rotation and W/H
 * scale controls.
 */
@Composable
fun SymbolControlsCard(
    waypoint: Waypoint,
    layers: List<DrawingLayer>,
    crosshairTargetLat: Double,
    crosshairTargetLng: Double,
    store: WaypointStore,
    onDismiss: () -> Unit,
    modifier: Modifier = Modifier
) {
    var showDeleteConfirm by remember { mutableStateOf(false) }
    var editMode by remember(waypoint.id) { mutableStateOf<SymbolEditorMode?>(null) }

    Column(
        modifier = modifier
            .shadow(elevation = 10.dp, shape = RoundedCornerShape(16.dp))
            .clip(RoundedCornerShape(16.dp))
            .background(Color(0xEE1C1C1E))
            .padding(horizontal = 12.dp, vertical = 10.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Header(
            waypoint = waypoint,
            onDismiss = onDismiss,
            onTitleClick = when (waypoint.kind) {
                WaypointKind.Generic -> null
                is WaypointKind.Military -> { { editMode = SymbolEditorMode.MILITARY } }
                is WaypointKind.ControlMeasure -> { { editMode = SymbolEditorMode.TASK } }
            }
        )

        when (waypoint.kind) {
            WaypointKind.Generic -> Unit
            is WaypointKind.ControlMeasure -> ControlMeasureControls(
                waypoint = waypoint,
                onWaypointChange = store::update,
                onWaypointChangeDraft = store::updateNoUndo
            )
            is WaypointKind.Military -> Unit
        }

        LayerSelectorButton(
            layers = layers,
            selectedLayerId = waypoint.layerId,
            onLayerSelected = { layerId ->
                store.update(waypoint.copy(layerId = layerId))
            },
            modifier = Modifier.fillMaxWidth()
        )

        ActionRow(
            waypoint = waypoint,
            crosshairTargetLat = crosshairTargetLat,
            crosshairTargetLng = crosshairTargetLng,
            store = store,
            onDelete = { showDeleteConfirm = true }
        )
    }

    if (showDeleteConfirm) {
        AlertDialog(
            onDismissRequest = { showDeleteConfirm = false },
            title = { Text("Delete symbol?") },
            text = { Text("This will permanently remove \"${waypoint.name}\".") },
            confirmButton = {
                TextButton(onClick = {
                    store.remove(waypoint)
                    showDeleteConfirm = false
                    onDismiss()
                }) {
                    Text("Delete", color = Color(0xFFFF3B30))
                }
            },
            dismissButton = {
                TextButton(onClick = { showDeleteConfirm = false }) { Text("Cancel") }
            }
        )
    }

    editMode?.let { mode ->
        SymbolEditorDialog(
            mode = mode,
            initialKind = waypoint.kind,
            initialName = waypoint.name,
            crosshairLat = null,
            crosshairLng = null,
            title = when (mode) {
                SymbolEditorMode.MILITARY -> "Change Military Unit"
                SymbolEditorMode.TASK -> "Change Tactical Task"
            },
            actionLabel = "Save",
            fullScreen = mode != SymbolEditorMode.TASK,
            onDismiss = { editMode = null },
            onConfirm = { name, kind ->
                store.update(waypoint.copy(name = name, kind = kind))
                editMode = null
            }
        )
    }
}

@Composable
private fun Header(
    waypoint: Waypoint,
    onDismiss: () -> Unit,
    onTitleClick: (() -> Unit)?
) {
    Row(verticalAlignment = Alignment.CenterVertically) {
        Box(
            modifier = Modifier
                .size(28.dp)
                .clip(RoundedCornerShape(6.dp))
                .background(Color.White),
            contentAlignment = Alignment.Center
        ) {
            Icon(
                if (waypoint.kind is WaypointKind.ControlMeasure) Icons.Default.Flag else Icons.Default.GpsFixed,
                contentDescription = null,
                tint = if (waypoint.kind == WaypointKind.Generic) Color(0xFFB48800) else Color.Black,
                modifier = Modifier.size(20.dp)
            )
        }
        Spacer(Modifier.size(10.dp))
        Column(
            modifier = Modifier
                .weight(1f)
                .then(if (onTitleClick != null) Modifier.clickable { onTitleClick() } else Modifier)
        ) {
            Text(
                waypoint.name,
                color = Color.White,
                fontSize = 14.sp,
                fontWeight = FontWeight.SemiBold,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis
            )
            Text(
                waypoint.kind.categoryDisplayName,
                color = Color.White.copy(alpha = 0.58f),
                fontSize = 11.sp,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis
            )
        }
        IconButton(onClick = onDismiss, modifier = Modifier.size(28.dp)) {
            Icon(
                Icons.Default.Close,
                contentDescription = "Close symbol controls",
                tint = Color.White.copy(alpha = 0.6f)
            )
        }
    }
}

/// Five-swatch colour picker for the task graphic. Black is the default;
/// the others follow the APP-6 affiliation palette (blue = friendly,
/// red = hostile, green = neutral, yellow = unknown). Mirrors iOS's
/// SymbolControlsCard colour row.
@Composable
private fun ColorRow(
    selected: TaskColor,
    onSelect: (TaskColor) -> Unit
) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Text(
            "Colour",
            color = Color.White.copy(alpha = 0.72f),
            fontSize = 12.sp,
            modifier = Modifier.weight(0.75f)
        )
        Row(
            horizontalArrangement = Arrangement.spacedBy(10.dp),
            modifier = Modifier.weight(2.45f)
        ) {
            TaskColor.entries.forEach { tc ->
                Box(
                    modifier = Modifier
                        .size(28.dp)
                        .clip(CircleShape)
                        .background(Color(tc.argb))
                        // White hairline so black/dark swatches read on the
                        // dark card; accent ring marks the current selection.
                        .border(
                            width = if (tc == selected) 3.dp else 1.dp,
                            color = if (tc == selected) Color(0xFF0A84FF) else Color.White.copy(alpha = 0.7f),
                            shape = CircleShape
                        )
                        .clickable { onSelect(tc) }
                )
            }
        }
    }
}

@Composable
private fun ControlMeasureControls(
    waypoint: Waypoint,
    onWaypointChange: (Waypoint) -> Unit,
    onWaypointChangeDraft: (Waypoint) -> Unit = onWaypointChange,
) {
    Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
        ColorRow(
            selected = waypoint.taskColor,
            onSelect = { onWaypointChange(waypoint.copy(taskColor = it)) }
        )
        SliderRow(
            title = "Rotation",
            value = waypoint.rotation.toFloat(),
            valueLabel = "${waypoint.rotation.toInt()}°",
            range = 0f..360f,
            onChange = { onWaypointChangeDraft(waypoint.copy(rotation = it.toDouble())) },
            onCommit = { onWaypointChange(waypoint.copy(rotation = it.toDouble())) },
            onReset = { onWaypointChange(waypoint.copy(rotation = 0.0)) }
        )
        SliderRow(
            title = "Width",
            value = waypoint.scaleX.toFloat(),
            valueLabel = "%.2fx".format(waypoint.scaleX),
            range = 0.15f..6f,
            onChange = { onWaypointChangeDraft(waypoint.copy(scaleX = it.toDouble())) },
            onCommit = { onWaypointChange(waypoint.copy(scaleX = it.toDouble())) },
            onReset = { onWaypointChange(waypoint.copy(scaleX = 1.0)) }
        )
        SliderRow(
            title = "Height",
            value = waypoint.scaleY.toFloat(),
            valueLabel = "%.2fx".format(waypoint.scaleY),
            range = 0.15f..6f,
            onChange = { onWaypointChangeDraft(waypoint.copy(scaleY = it.toDouble())) },
            onCommit = { onWaypointChange(waypoint.copy(scaleY = it.toDouble())) },
            onReset = { onWaypointChange(waypoint.copy(scaleY = 1.0)) }
        )
    }
}

@Composable
private fun ActionRow(
    waypoint: Waypoint,
    crosshairTargetLat: Double,
    crosshairTargetLng: Double,
    store: WaypointStore,
    onDelete: () -> Unit
) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Button(
            onClick = {
                store.update(waypoint.copy(
                    latitude = crosshairTargetLat,
                    longitude = crosshairTargetLng
                ))
            },
            modifier = Modifier.weight(1f).height(36.dp),
            colors = ButtonDefaults.buttonColors(
                containerColor = Color(0xFF0A84FF).copy(alpha = 0.85f),
                contentColor = Color.White
            ),
            shape = RoundedCornerShape(8.dp),
            contentPadding = PaddingValues(horizontal = 12.dp, vertical = 6.dp)
        ) {
            Icon(Icons.Default.GpsFixed, contentDescription = null, modifier = Modifier.size(14.dp))
            Spacer(Modifier.size(6.dp))
            Text("Move to Crosshair", fontSize = 13.sp, fontWeight = FontWeight.SemiBold)
        }
        Button(
            onClick = onDelete,
            modifier = Modifier.size(width = 44.dp, height = 36.dp),
            colors = ButtonDefaults.buttonColors(
                containerColor = Color(0xFFFF3B30).copy(alpha = 0.85f),
                contentColor = Color.White
            ),
            shape = RoundedCornerShape(8.dp),
            contentPadding = PaddingValues(0.dp)
        ) {
            Icon(Icons.Default.Delete, contentDescription = "Delete symbol", modifier = Modifier.size(16.dp))
        }
    }
}

@Composable
private fun SliderRow(
    title: String,
    value: Float,
    valueLabel: String,
    range: ClosedFloatingPointRange<Float>,
    onChange: (Float) -> Unit,
    onCommit: (Float) -> Unit = onChange,
    onReset: () -> Unit
) {
    var latestValue by remember(value) { mutableStateOf(value) }
    Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
        Text(title, color = Color.White.copy(alpha = 0.72f), fontSize = 12.sp, modifier = Modifier.weight(0.75f))
        Slider(
            value = latestValue.coerceIn(range.start, range.endInclusive),
            onValueChange = { latestValue = it; onChange(it) },
            onValueChangeFinished = { onCommit(latestValue) },
            valueRange = range,
            modifier = Modifier.weight(1.8f)
        )
        Text(valueLabel, color = Color.White.copy(alpha = 0.76f), fontSize = 11.sp, modifier = Modifier.weight(0.65f))
        TextButton(onClick = onReset, contentPadding = PaddingValues(horizontal = 4.dp)) {
            Text("Reset", fontSize = 11.sp)
        }
    }
}
